pkg/kn/flags/bool.go

```diff
@@ -0,0 +1,79 @@
+// Copyright © 2018 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package flags
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/pflag"
+)
+
+var negPrefix = "no-"
+
+// AddBothBoolFlags adds the given flag in both `--foo` and `--no-foo` variants.
+// If you do this, make sure you call ReconcileBoolFlags later to catch errors and
+// set the relationship between the flag values.
+func AddBothBoolFlags(f *pflag.FlagSet, p *bool, name, short string, value bool, usage string) {
+
+	negativeName := negPrefix + name
+
+	f.BoolVarP(p, name, short, value, usage)
+	f.Bool(negativeName, !value, "do not "+usage)
+
+	if value {
+		err := f.MarkHidden(name)
```

Comment:
maybe just not add it if the option already corresponds to the default value? Not sure, but that might help to catch errors as calling the hidden command will result in a no-op which might be hiding errors.

On the other hand, it's good to have those in case we want to change the default value later to the opposites. That would be for sure a backwards-incompatible change, but a user could already safeguard against such a change by always using the appropriate option.

So I'm for to keep it like it is (so ignore my comment ;-)

Reply:
btw, and somewhat unrelated, that's my general take on default values in general (included in the book I wrote). Opinionated, so I don't expect that everyone agrees on this ;-)

> Default values make life easier, as they take away the burden of selecting a value for a configuration parameter you might not even know exists. They also play a significant role in the _convention over configuration_ paradigm. However, defaults are not always a good idea. Sometimes they might even be an antipattern for an evolving application.
>
> This is because _changing_ default values retrospectively is a difficult task.
First, changing default values means replacing them within the code, which requires a rebuild. Second, people relying on defaults (either by convention or consciously) will always be surprised when a default value changes. We have to communicate the change, and the user of such an application probably has to modify the calling code as well.
> 
> Changes in default values, however, often make sense, because it is hard to get default values right from the very beginning. It's essential that we consider a change in a default values as a _major change_, and if semantic versioning is in use, such a modification justifies a bump in the major version number. If unsatisfied with a given default value, it is often better to remove the default altogether and throw an error if the user does not provide a configuration value. This will at least break the application early and prominently instead of it doing something different and unexpected silently.
> 
> Considering all these issues, it is often the best solution to _avoid default values_ from the very beginning if you cannot be 90% sure that a reasonable default will last for a long time. Passwords or database connection parameters are good candidates for not providing default values, as they depend highly on the environment and often cannot be reliably predicted. Also, if we do not use default values, the configuration information has to be provided explicitly, which serves as documentation too.

Reply:
Yeah, I went back and forth on the whole "do you have a flag that basically is a noop to do the default thing" issue, but was settled by the case where it's also a config file-settable variable, which effectively makes a context-dependent default, and you definitely want both flags. So leaving them in.